pkgs/development/libraries/raylib/default.nix

```diff
@@ -0,0 +1,54 @@
+{ stdenv, lib, fetchFromGitHub, fetchpatch, cmake,
+  mesa, libGLU, glfw,
+  libX11, libXi, libXcursor, libXrandr, libXinerama,
+  alsaSupport ? stdenv.hostPlatform.isLinux, alsaLib,
+  pulseSupport ? stdenv.hostPlatform.isLinux, libpulseaudio,
+  includeEverything ? true
+}:
+
+stdenv.mkDerivation rec {
+  pname = "raylib";
+  version = "3.5.0";
+
+  src = fetchFromGitHub {
+    owner = "raysan5";
+    repo = pname;
+    rev = version;
+    sha256 = "0syvd5js1lbx3g4cddwwncqg95l6hb3fdz5nsh5pqy7fr6v84kwj";
+  };
+
+  patches = [
+    # fixes examples not compiling in 3.5.0
+    (fetchpatch {
+      url = "https://patch-diff.githubusercontent.com/raw/raysan5/raylib/pull/1470.patch";
+      sha256 = "1ff5l839wl8dxwrs2bwky7kqa8kk9qmsflg31sk5vbil68dzbzg0";
+    })
+  ];
+
+  nativeBuildInputs = [ cmake ];
+  buildInputs = [
+    mesa libGLU glfw libX11 libXi libXcursor libXrandr libXinerama
+  ] ++ lib.optional alsaSupport alsaLib
+    ++ lib.optional pulseSupport libpulseaudio;
+
+  # https://github.com/raysan5/raylib/wiki/CMake-Build-Options
+  cmakeFlags = [
+    "-DUSE_EXTERNAL_GLFW=ON"
+    "-DSHARED=ON"
+    "-DBUILD_EXAMPLES=OFF"
```

Comment:
It looks like the examples dont end up in the final output directory, only remaining in the build dir.
Is there any value in re-enabling them?

Reply:
Well, let me know if I understood it correctly.

The examples generate executable programs and they work, but these generated executable files are not pushed in the final dir?

If this is the case, there is no problem. After all, these examples are merely for testing purposes, and there is no serious reason to save the results of a test. (Except maybe if some fun and useful thing comes with them, like the famous `glxgears`.) 

On the other hand, it is useful to save the source code of these examples. Put these codes in a directory like `$out/share/doc/${pname}/examples/`

Reply:
Thanks for the suggestion. I appreciate you taking the time to hand-hold me through this.

Looking at the docs, the best phase I could find was the `postInstall` phase.
I copy all the example source to $out/share/doc/${pname}/examples
Then cleanup the root directory which has some CMakeLists.txt, Makefile, *.bat, and other files which shouldn't be relevant.

I looked through the nixpkgs repo and couldn't find any other libraries that build nor copy the examples (SDL/2, cegui, mygui, etc).
(As an aside, I also was unable to get github search to work. I have a feeling the repo is too big)

I think it's good to copy the examples, as you say.
I'm thinking it might be good to build and copy the example binaries too.
Thoughts?
Would this be in `$out/share/bin`?

The only issue is, you would double up on all the assets, as they are in the example files next to the relevant source.
Then CMake moves them to the collective location with the output binaries.
Perhaps this is just more trouble than it's worth and what we've got now is sufficient?

Reply:
Yes, the Nixpkgs is a huge monster nowadays. It has almost 2G (and `pkgs/` has 134M) according to `du -sh`.
But you can search the local copy with an utility like ripgrep.

About the installation: Well, I am thinking a bit about it.
The example binaries are not so useful. The end users can copy the source codes of these examples and compile them by themselves.
If the CMake scripts of Raylib does not install these example binaries, then leave them out - forget them.

My suggestion is: 

- Don't install the source code of the examples - yes, forget them;
- If the CMake scripts don't install the examples binaries, forget them too.

Yes, it is much squeeze for little juice.

Reply:
Sorry, when you say "enable the tests", do you mean build the examples but not package them?

Reply:
Yes. Use the examples in a `checkPhase`, but don't install them.

Reply:
The examples are graphical and require the user to press ESC to exit, they aren't simple CLI programs which quit.
I wouldn't be able to execute them, which I think is what you're recommending I do.
I think this is where our cross-communication is coming in?

I can re-enable them as part of the build, but I think there would be little value there if they're not executed.

Reply:
Well, the mere fact they are compiling without errors is a successful test, even if they are interactive.

But you are right, there is little value in these tests, at least in our current framework. Disable the examples.

(NixOS has a unit test framework, based on virtual machines, and the interactive programs can be controlled on it. But forget this for now.)